### ISSUE TITLE
Reduce memory usage on startup

### ISSUE THREAD

**Issue by bpringe** [clojure-lsp/clojure-lsp#229]
I want to open this to start/track a conversation about reducing clojure-lsp's memory usage.

I noticed the jvm opts are `:jvm-opts ^:replace ["-Xmx2g" "-server"]`. Is 2g necessary as a max? Potentially if you have n instances running they could use up to `2 x n` gigs of memory. Have lower limits been tried, and what are the results?

If there is an issue with a lower max heap size, it should be possible to make clojure-lsp hold less data in memory at any given time with some tweaks.

Since people are often also running REPLs in their projects, this memory consumption is added on top of the jvm instance for the REPL. I suppose this may be some people's arguments for driving all features through the REPL as opposed to a static analyzer. 

The REPL is already consuming a certain amount of memory, so starting another jvm for features the REPL can offer seems like unnecessary added memory consumption. I'm just throwing an argument against clojure-lsp as it stands now, but obviously I think this is a great project and it can be improved in this regard.

**Comment by ericdallo** [comment]
It's a good question, I know @snoe increased the mem last time from 1 GB -> 2GB but I don't know why or if it was related to clj-kondo.
I think we can try reducing it indeed, I'm not sure clojure-lsp needs that amount of memory just for it's `db` atom.
We can test it reducing to 1GB, WDYT?

**Comment by ericdallo** [comment]
I know that at that time we had a memory leak issue with clj-kondo that was already solved

**Comment by bpringe** [comment]
We'll see if @snoe has some insights, as I'm sure he had a reason for the increase, but I think testing lower limits would be a good start. (

**Comment by ericdallo** [comment]
I found this: https://github.com/clojure-lsp/clojure-lsp/commit/b241b795f45210dcbb2d15c2d04d9cd2616a07d6

**Comment by bpringe** [comment]
I see, so it seems the memory issue was fixed or improved in kondo, but the heap size was still increase in lsp, maybe just to be safe?

On another note, VisualVM may be a good tool for profiling clojure-lsp. I've seen idle instances of clojure-lsp running at 1.7G, which seems way too high considering it's dealing with text files that combined should never reach that size, and also it shouldn't be holding that much in memory anyway.

**Comment by ericdallo** [comment]
Yes, we probably are missing some improvements in performance, specially on this startup function: https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/crawler.clj#L145-L185

**Comment by bpringe** [comment]
I've seen some cool usages of transducers in clojure-lsp. I admittedly haven't used them much before, but it seems they may help with memory usage? Maybe we could replace more code with transducers.

**Comment by ericdallo** [comment]
Neither I, it's a clojure pattern but not necessarily will improve performance, for clojure-lsp case, I think we should use more async API on startup.
Also, something that will surely improve the performance is to finish this PR: https://github.com/clojure-lsp/clojure-lsp/pull/176

**Comment by bpringe** [comment]
I think transducers can improve performance and reduce memory usage. See [here](https://clojure.org/guides/faq#transducers_vs_seqs).

>  Source collection iteration - when used on reducible inputs (collections and other things), avoid creating an unnecessary input collection sequence - helps memory and time.

> Intermediate sequences and cached values - as the transformation happens in a single pass, you remove all intermediate sequence and cached value creation - again, helps memory and time. The combination of the prior item and this one will start to win big as the size of the input collection or number of transformations goes up (but for small numbers of either, chunked sequences can be surprisingly fast and will compete).

**Comment by bpringe** [comment]
Though, idk if there are places where this is needed in clojure-lsp, just speaking generally.

**Comment by ericdallo** [comment]
Yes, I agree is good point, I will take a look later for this kind of improvement, thanks!

**Comment by ericdallo** [comment]
Just a note: [This PR](https://github.com/clojure-lsp/clojure-lsp/pull/230) does not improve memory but runs things on other threads, which may affect memory.

**Comment by ericdallo** [comment]
I did some tests on big projects, and I could not reduce the 2GB xmx, it seems that clojure-lsp just needs a high amount of memory on the startup to crawl jars, files and the call clj-kondo on the classpath, after that, it uses way less memory running the entire application.
WDYT @snoe, can we think in anything to improve that?

**Comment by bpringe** [comment]
What happens if you reduce the max? Do you get OOM errors? I think there are also ways to tweak the JVM's garbage collection to run differently, so toying with those settings may provide some improvement.

**Comment by ericdallo** [comment]
Yes, for big projects you get OOM errors.
Yeah I think, but I'm not familiar with tweaking those settings

**Comment by FongHou** [comment]
I found two things helped reducing memory usage..

**First**, I  added (System/gc) in handlers.clj initialize. Here is my local repo diff. 

```
diff --git a/src/clojure_lsp/handlers.clj b/src/clojure_lsp/handlers.clj
index 64da516..75fc858 100644
--- a/src/clojure_lsp/handlers.clj
+++ b/src/clojure_lsp/handlers.clj
@@ -81,27 +79,33 @@
 (defn initialize [project-root client-capabilities client-settings]
   (when project-root
     (let [project-settings (crawler/find-project-settings project-root)]
       (swap! db/db assoc
              :project-root project-root
              :project-settings project-settings
              :client-settings client-settings
              :settings (-> (merge client-settings project-settings)
                            (update :cljfmt cljfmt.main/merge-default-options))
              :client-capabilities client-capabilities))
     (let [file-envs (crawler/determine-dependencies project-root)]
       (swap! db/db assoc
              :file-envs file-envs
-             :project-aliases (apply merge (map (comp :aliases val) file-envs))))))
+             :project-aliases (apply merge (map (comp :aliases val) file-envs))))
+    (let [before-gc (.totalMemory (Runtime/getRuntime))]
+      (System/gc)
+      (log/info "GC freed up memory "
+                (/ (- before-gc (.totalMemory (Runtime/getRuntime)))
+                   (* 1024 1024))
+                " MB"))))
```

As screenshot shows, JVM heap size falled off a cliff and return freed memory back to OS.

![image](https://user-images.githubusercontent.com/13973254/103588236-c0c3bd80-4eb6-11eb-878f-f16842aee739.png)

clojure-lsp.out shows total memory freed up by (System/gc) call.

```
INFO  clojure-lsp.crawler: starting clj-kondo project classpath scan (this takes awhile)
INFO  clojure-lsp.crawler: clj-kondo scanned project classpath in 27178ms
INFO  clojure-lsp.handlers: GC freed up memory  1515  MB
```

**Second**, I found that lein classpath includes lots of jars that are irrelevent to the project. Tuning number of jars from that command can make big different in memory usage and project initial scanning time. I put the second command in my .lsp/config.edn classpath-cmd. It cut down memory usage and scanning time significantly. 

```
❯ lein classpath | sed -e 's/:/\n/g' | wc -l
485

❯ lein with-profile dev classpath | sed -e 's/:/\n/g' | wc -l
115
```

**Comment by ericdallo** [comment]
@FongHou thanks for helping on that!
About the first, could you check again without the `System/gc`? In my tests the memory already drops on that exact point of code, after determining the dependencies.

About the second, is this safe? we always want to use the dev profile and remove a lot of jars? What if user doesn't have dev profile? What if user really wants all jars on its project? We could have defaults for that, but I just want to bring these questions

**Comment by FongHou** [comment]
Hi @ericdallo,

About the first, Full Stop-the-World GC may eventually happen, but is not deterministic. I tested a few times on my laptop without (System/gc) call, it didn't happen even for tens miniute (I guess it depends JVM version and OS). Explicit System/gc call (though doesn't guarantee FSTW GC), but it's much more deterministic.

About the second, dev profile is one of a few lein buildin profiles. You don't need to explicitly define it. It'll always include source-pathes and dependent jars listed in project.clj.

**Comment by ericdallo** [comment]
> About the first, Full Stop-the-World GC may eventually happen, but is not deterministic. I tested a few times on my laptop without (System/gc) call, it didn't happen even for tens miniute (I guess it depends JVM version and OS). Explicit System/gc call (though doesn't guarantee FSTW GC), but it's much more deterministic.

Ok, it makes sense, but is calling a Full GC a good approach here? we may reduce memory usage, but we may slow down until GC is done, WDYT?

> About the second, dev profile is one of a few lein buildin profiles. You don't need to explicitly define it. It'll always include source-pathes and dependent jars listed in project.clj.

I see. that's something we can try, making a setting like `classpath-scan-profile` which if not provided it will use `dev` or if provided empty, it will use no profile

**Comment by FongHou** [comment]
System/gc shouldn't be used in hot code path in genenal. I think using it in handlers.clj initialize is fine becaut that function only run once for the whole life time of JVM.

About classpath tuning, maybe just put some tips in README. I'm not sure it's worth it in the code. My ~/.lein/profiles.clj pulled in lots of repl/user profile jars, which lein classpath by default that's why it made a big different. clj deps.edn may behave differently.

**Comment by ericdallo** [comment]
> About classpath tuning, maybe just put some tips in README. I'm not sure it's worth it in the code. My ~/.lein/profiles.clj pulled in lots of repl/user profile jars, which lein classpath includes by default that's why it made a big different. clj deps.edn may 
behave differently.
 
 I like tuning defaults, where user can just install and everything works great. If we confirm that will reduce a lot jars most projects we can add this to code with a flag to customize it.
 I will try to check the huge projects from my company tomorrow morning and see how big is the difference
 
 > clj deps.edn may behave differently.
 
Yeah, probably different, lein classpath for `clojure-lsp`(it uses deps.edn) returns 66 jars while with dev profile return 64.

**Comment by bpringe** [comment]
Awesome investigating, @FongHou. @ericdallo, if you think the kondo PR mentioned will improve performance, perhaps we should wait for those changes to re-assess what else should be done after that. I assume it could/should still be improved even after those changes.

Regarding GC, calling System/gc might be a good idea, but I think there could be some easy wins in code refactoring that should be explored first, like eliminating intermediary collections using transducers where possible, and maybe even replacing some recursion with an iterative approach to save on the memory that gets built up in the stack (I'm not sure if or where this could be an issue in clojure-lsp, at the moment, just something to look into).

**Comment by bpringe** [comment]
I'm going to look into / play around with this now. I've not done a whole lot of jvm/Clojure performance tuning, but this article is a great read if interested: https://tech.redplanetlabs.com/2020/09/02/clojure-faster/

**Comment by ericdallo** [comment]
Really good article @bpringe, I did some performance/benchmark checks with the `criterium` lib, and it seems what is not so fast it's this function that is called for each file of each jar: https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/feature/references.clj#L43
Improving the performance could help reduce the memory

**Comment by ericdallo** [comment]
Briefly, the function that takes more time is [parser/find-usages](https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/parser.clj#L1174) which is the function that parse the code block and build the main core of our db.
As I said, I believe after #176, we'll probably don't need the entire `parser.clj` ns, and relies 100% on `clj-kondo` to parse the code and get the usages to save on DB.
Meanwhile, we could try to improve the performance of the `parser.clj` namespace.

**Comment by bpringe** [comment]
@ericdallo I see, thanks for the info. A friend of mine helped me to realize that performance (speed) and memory usage are not really the same thing (though of course somewhat related), so I want to shift my focus to memory usage instead, which may be a bit more elusive, but I've been dealing with another issue for some time now and haven't gotten back to this yet.

Anyway, my thoughts are that to figure out where memory issues are happening, besides just looking at the code and making educated guesses, is to drastically lower the max heap size to cause the server to crash from an OOM error, and I _think_ the stack trace(s) should give us a good idea of where the problem(s) are, provided they're consistent.

**Comment by ericdallo** [comment]
Hey @bpringe  good news, after #261 we probably improved memory management since we delegate to clj-kondo to analyze that (which should analyze in a smarter way than we were analyzing before), I will try to make some tests this weekend and check if we can lower the 2GB flag

**Comment by ericdallo** [comment]
I made some performance tests and AFAICS the #261 improved a lot the memory usage on startup, and after that it's even low as it was before. Therefore I decreased the memory usage from 2GB -> 1.5GB https://github.com/clojure-lsp/clojure-lsp/commit/6ebf0c4931563b026f40384b218caf3666825832

**Comment by bpringe** [comment]
Just did a little test on a small project. When I opened it and lsp had already crawled it, the memory usage went up to and stayed around ~370 MiB. When I deleted the .lsp and .clj-kondo directories and reopened the project, lsp memory went up to and stayed around ~1.3 GiB.

It seems to hog a lot of memory for no good reason. Maybe the GC doesn't kick in because it's not reaching the max heap size (1.5 GB) though. A while back when I used VisualVM to analyze the server, I used the GC button while mem usage was hover over 1 GB and it dropped way lower. I think this coincides with the discussion about calling `System/gc` above. (Not saying that this should or shouldn't be done, just pointing it out.)

**Comment by FongHou** [comment]
One caveat with manual GC is that JVM may or maynot release memory back to OS immediately. It varies depending on JVM version and OS as disicussed here https://stackoverflow.com/questions/30458195/does-gc-release-back-memory-to-os.

I've been using "java -Xmx1g -XX:+UseG1GC -XX:MaxHeapFreeRatio=30 -XX:MinHeapFreeRatio=10" with a local (System/gc) build, git diff like below

```
diff --git a/src/clojure_lsp/crawler.clj b/src/clojure_lsp/crawler.clj
index 42515f9..841cfcd 100644
--- a/src/clojure_lsp/crawler.clj
+++ b/src/clojure_lsp/crawler.clj
@@ -195,7 +195,7 @@
         source-paths (get settings :source-paths)]
     (analyze-classpath root-path source-paths settings)
     (analyze-paths source-paths false)
-    nil))
+    (System/gc)))
```

Ran on a project with hundreds external jars to scan fresh (rm -rf .clj-kondo/.cache .lsp/sqlite.db before). RSS was hovering around 1.5GB (0.5GB is non-heap overhead, no OOME even with -Xmx1g), and immediately dropped down to 560MB right after clojure-lsp finished the scan. Here is a screenshot from my ubuntu wsl2 running java 8.

![image](https://user-images.githubusercontent.com/13973254/106202431-ec6d5700-6187-11eb-87ac-25e9c1701ad4.png)

**Comment by FongHou** [comment]
btw, not noticeable pause time of (System/gc) and how long it takes to complete the scan.

**Comment by ericdallo** [comment]
Yeah, this is something to consider, indeed the memory should have a huge drop after the startup, @snoe do you see any major issues calling System/gc after the startup classpath analysis?

**Comment by ericdallo** [comment]
Hey @bpringe @FongHou I pushed the manual gc: https://github.com/clojure-lsp/clojure-lsp/commit/16af95c06419940dc9d5eb6626ad735dc5c072c7
It was pretty fast (few ms) even in a big project, could you test it later?

**Comment by bpringe** [comment]
I tested version `2021.02.02-14.02.23` with a simple project, first removing the .lsp and .clj-kondo files, and I still see just as high of idle memory after startup (1.2+ GB) as before this version. I'm only checking System Monitor though, but I'd expect it to show if mem was dropped after the GC.

**Comment by ericdallo** [comment]
Hum, it really drop the mem after the GC in my tests 🤔

**Comment by ericdallo** [comment]
@bpringe we just call gc manually on batch analysis, Could you check if your project have more than 50 files on classpath (check the logs)?

**Comment by bpringe** [comment]
```
2021-02-03T00:29:43.977Z brandon-desktop WARN [clojure-lsp.db:45] - Could not load db path to '/home/brandon/development/clojure-test/.lsp/sqlite.db': '/home/brandon/development/clojure-test/.lsp' does not exist
2021-02-03T00:29:43.996Z brandon-desktop INFO [clojure-lsp.crawler:111] - Analyzing 33 paths with clj-kondo with batch size of 1 ...
2021-02-03T00:29:58.906Z brandon-desktop INFO [clojure-lsp.crawler:165] - Paths analyzed, took 14.909956 secs. Caching for next startups...
2021-02-03T00:29:59.058Z brandon-desktop INFO [clojure-lsp.crawler:194] - Manual GC after classpath scan took 0.12044575 seconds
```
Only 33 paths it seems, but still looks like it does the manual GC

**Comment by ericdallo** [comment]
Oh, the batch check must not be working, anyway, is not guaranteed that JVM will run the GC, I don't know if that could be the issue :/
I think refactoring this code using transducer would probably remove the necessity of this GC.

**Comment by bpringe** [comment]
Could also be differences in JDK versions :man_shrugging:

**Comment by ericdallo** [comment]
So, I didn't hear any complaints about memory after the GraalVM migration + performance improvements done later, I'm comfortable with closing this, we can re-open it if makes sense later. WDYT @bpringe?

**Comment by bpringe** [comment]
That sounds good to me.
